govuk-one-login/authentication-frontend
# BAU: Make docker-compose listening ports easily editable

## What

### make dev docker exposed ports modifiable

Rather than the ports being hardcoded in `docker-compose.yml`, instead set them via environment variables.

This means people can change the ports locally in a non-committed file in the case of a clash (an existing service listening on the same port)

### YAML formatting in `docker-compose.yml`

Rather than using the `list` format for environment variables, instead define them as key:value pairs.

This makes the key and the value visually distinct with syntax highlighting enabled.

### Add checks for required environment variables in `docker compose`

Rather than just showing a warning for missing variables, instead error when required ones are not defined.

This means that the frontend can only be started when all the environment variables are set, reducing confusion
in the case that a required one is not set, and it's not obvious that a missing envar is the cause.

### Add tests for required node environment variables

Some values are not provided in the samples, and need to be added. Weird stuff happens when they're unset.

As we know that missing these will cause problems, we can check them in advance.

Additionally, this gives us a good place to test if the `$ORCH_TO_AUTH_SIGNING_KEY` value is correctly formatted - it cannot be set with `\n`, but has to have actual newline characters to be processed correctly.

## How to review


> NB: ensure that you keep a copy of your `.env` file, so it's not lost forever!

1. Ensure the readme changes are sensible / readable.
1. remove your `.env` file
1. run `startup.sh`: note that relevant errors are thrown about missing envars.
1. restore the `.env` file
1. run `startup.sh`: note that the errors go away and the startup completes correctly
1. run `docker compose up`: note that the startup fails due to missing AWS variables. note that the error message explains how this should be set
1. change one of the `DOCKER_.*_PORT` variables in `.env`
1. run `startup.sh`: note that the service whos port you changed is now listening on the new port
1. ensure that the service is accessible on that new port
